aten/src/ATen/test/MaybeOwned_test.cpp

```diff
@@ -0,0 +1,297 @@
+#include <gtest/gtest.h>
+
+#include <ATen/NativeFunctions.h>
+#include <ATen/Tensor.h>
+#include <c10/util/intrusive_ptr.h>
+#include <c10/util/MaybeOwned.h>
+
+#include <memory>
+#include <string>
+
+namespace {
+
+template <typename T>
+using MaybeOwned = c10::MaybeOwned<T>;
+
+using at::Tensor;
+
+struct MyString : public c10::intrusive_ptr_target, public std::string {
+  using std::string::string;
+};
+
+template <typename T>
+class MaybeOwnedTest : public ::testing::Test {
+ public:
+  using MaybeOwned = c10::MaybeOwned<T>;
+
+  T borrowFrom;
+  T ownCopy;
+  T ownCopy2;
+  MaybeOwned borrowed;
+  MaybeOwned owned;
+  MaybeOwned owned2;
+
+ protected:
+  void SetUp() override; // defined below helpers
+  void TearDown() override {
+    // Release everything to try to trigger ASAN violations in the
+    // test that broke things.
+    borrowFrom = T();
+    ownCopy = T();
+    ownCopy2 = T();
+
+    borrowed = MaybeOwned();
+    owned = MaybeOwned();
+    owned2 = MaybeOwned();
+  }
+
+};
+
+
+//////////////////// Helpers that differ per tested type. ////////////////////
+
+template <typename T>
+T getSampleValue();
+
+template <typename T>
+T getSampleValue2();
+
+template <typename T>
+bool equal(const T&, const T&);
+
+template <typename T>
+void assertBorrow(const MaybeOwned<T>&, const T&);
+
+template <typename T>
+void assertOwn(const MaybeOwned<T>&, const T&, size_t useCount = 2);
```

Comment:
Personally I'd omit the default param here, I think it just adds indirection when reading the callsites, but not a strong opinion